pkg/apis/options/legacy_options.go

```diff
@@ -403,3 +424,192 @@ func getXAuthRequestAccessTokenHeader() Header {
+	flagSet.StringSlice("allowed-group", []string{}, "restrict logins to members of this group (may be given multiple times)")
+
+	return flagSet
+}
+
+func (l *LegacyProvider) convert() (Providers, error) {
+	providers := Providers{}
+
+	provider := Provider{
+		ClientID:          l.ClientID,
+		ClientSecret:      l.ClientSecret,
+		ClientSecretFile:  l.ClientSecretFile,
+		ProviderType:      l.ProviderType,
+		ProviderCAFiles:   l.ProviderCAFiles,
+		LoginURL:          l.LoginURL,
+		RedeemURL:         l.RedeemURL,
+		ProfileURL:        l.ProfileURL,
+		ProtectedResource: l.ProtectedResource,
+		ValidateURL:       l.ValidateURL,
+		Scope:             l.Scope,
+		Prompt:            l.Prompt,
+		ApprovalPrompt:    l.ApprovalPrompt,
+		AllowedGroups:     l.AllowedGroups,
+		AcrValues:         l.AcrValues,
+	}
+
+	// This part is out of the switch section for all providers that support OIDC
+	provider.OIDCConfig = OIDCOptions{
+		OIDCIssuerURL:                      l.OIDCIssuerURL,
+		InsecureOIDCAllowUnverifiedEmail:   l.InsecureOIDCAllowUnverifiedEmail,
+		InsecureOIDCSkipIssuerVerification: l.InsecureOIDCSkipIssuerVerification,
+		SkipOIDCDiscovery:                  l.SkipOIDCDiscovery,
+		OIDCJwksURL:                        l.OIDCJwksURL,
+		UserIDClaim:                        l.UserIDClaim,
+		OIDCEmailClaim:                     l.OIDCEmailClaim,
+		OIDCGroupsClaim:                    l.OIDCGroupsClaim,
+	}
+
+	// This part is out of the switch section because azure has a default tenant
+	// that needs to be added from legacy options
+	provider.AzureConfig = AzureOptions{
+		AzureTenant: l.AzureTenant,
+	}
+
+	switch provider.ProviderType {
+	case "github":
+		provider.GitHubConfig = GitHubOptions{
+			GitHubOrg:   l.GitHubOrg,
+			GitHubTeam:  l.GitHubTeam,
+			GitHubRepo:  l.GitHubRepo,
+			GitHubToken: l.GitHubToken,
+			GitHubUsers: l.GitHubUsers,
+		}
+	case "keycloak":
+		provider.KeycloakConfig = KeycloakOptions{
+			KeycloakGroups: l.KeycloakGroups,
+		}
+	case "gitlab":
+		provider.GitLabConfig = GitLabOptions{
+			GitLabGroup:    l.GitLabGroup,
+			GitlabProjects: l.GitlabProjects,
+		}
+	case "login.gov":
+		provider.LoginGovConfig = LoginGovOptions{
+			JWTKey:     l.JWTKey,
+			JWTKeyFile: l.JWTKeyFile,
+			PubJWKURL:  l.PubJWKURL,
+		}
+	case "bitbucket":
+		provider.BitbucketConfig = BitbucketOptions{
+			BitbucketTeam:       l.BitbucketTeam,
+			BitbucketRepository: l.BitbucketRepository,
+		}
+	case "google":
+		provider.GoogleConfig = GoogleOptions{
+			GoogleGroups:             l.GoogleGroups,
+			GoogleAdminEmail:         l.GoogleAdminEmail,
+			GoogleServiceAccountJSON: l.GoogleServiceAccountJSON,
+		}
+	}
+
+	if l.ProviderName != "" {
+		provider.ProviderID = l.ProviderName
+		provider.ProviderName = l.ProviderName
+	} else {
+		// TODO (yanasega): should set a better default id value
+		provider.ProviderID = l.ProviderType + "_" + l.ClientID
```

Comment:
I think the ID is probably ok as is, do we need this TODO? What other pattern would you suggest?

Reply:
I think the `_` might collide and cause ambiguity with allowed names in both those two fields its separating 🤔 

Maybe `=`, `/` or `:` as the separator?

Reply:
going with '='